src/react/hooks/internal/useStableCallback.ts

```diff
@@ -0,0 +1,24 @@
+import React from "rehackt";
+
+/**
+ * A hook that provides a stable reference to a function without suffering from
+ * stale closures. Useful to get a stable function reference for callback
+ * function options in other hooks. This avoids the need for the user to wrap
+ * each callback in a `useCallback`.
+ *
+ * @param unstableCallback - A callback function
+ */
+export function useStableCallback<TArgs extends unknown[], TReturn>(
+  unstableCallback: (...args: TArgs) => TReturn
+) {
+  const callbackRef = React.useRef(unstableCallback);
+
+  React.useLayoutEffect(() => {
+    callbackRef.current = unstableCallback;
+  });
+
+  return React.useCallback((...args: TArgs): TReturn => {
+    const fn = callbackRef.current;
```

Comment:
```suggestion
export function useStableCallback<T extends Function>(unstableCallback: T): T;
export function useStableCallback(unstableCallback: (...args: any[]) => any) {
  const callbackRef = React.useRef(unstableCallback);

  React.useLayoutEffect(() => {
    callbackRef.current = unstableCallback;
  });

  return React.useCallback((...args: any[]) => {
```

This would make the generic signature more in line with `useCallback` and would make it easier to use in `useLazyQuery`

Reply:
Nice trick! I didn't think about trying to function overload signature to work around the issue I was seeing. Thanks!

Reply:
[8a45baa](https://github.com/apollographql/apollo-client/pull/11403/commits/8a45baa546c4a3d0cf17c6eacb0a8c757a08867f)